lib/heuristics/periodic_heuristic.rb

```diff
@@ -464,83 +476,83 @@ def reorder_stops(vrp)
         end
       end
 
       index = removed_index
       while stops[index] && [:never_first, :always_middle].include?(stops[index][:requirement])
-        clean_stops(stops[index][:id], vehicle_id, true)
+        clean_stops(stops[index][:id], true)
         # index removed so no need to increment index
       end
     end
 
-    def clean_stops(service_id, vehicle_id, reaffect = false)
-      ### when allow_partial_assignment is false, removes all affected visits of [service_id] because we can not affect all visits ###
-      @candidate_routes[vehicle_id].collect{ |day, route_data|
-        remove_index = route_data[:stops].find_index{ |stop| stop[:id] == service_id }
-        next unless remove_index
-
-        used_point = route_data[:stops][remove_index][:point_id]
-        route_data[:stops].slice!(remove_index)
-        @services_data[service_id][:capacity].each{ |need, qty| route_data[:capacity_left][need] += qty }
+    def make_available(service_id)
+      @candidate_services_ids << service_id
+      @candidate_routes.each{ |_vehicle_id, all_routes|
+        all_routes.each{ |day, route_data|
+          next unless day <= @services_data[service_id][:raw].last_possible_days.first # first is for first visit
 
-        update_point_vehicle_and_days(used_point)
-        clean_position_dependent_services(vehicle_id, route_data[:stops], remove_index) unless route_data[:stops].empty?
-        route_data[:stops] = update_route(route_data, remove_index)
-        @services_data[service_id][:used_days] = []
-        @services_data[service_id][:used_vehicles] = []
-      }.compact.size
+          route_data[:available_ids] |= service_id
+        }
+      }
+      @services_assignment[service_id][:missing_visits] = @services_data[service_id][:raw].visits_number
+      @services_assignment[:unassigned_reasons] = []
+    end
 
-      if reaffect
-        @candidate_services_ids << service_id
-        @candidate_routes.each{ |_vehicle_id, all_routes|
-          all_routes.each{ |day, route_data|
-            next unless day <= @services_data[service_id][:raw].last_possible_days.first # first is for first visit
+    def clean_stops(service_id, reaffect = false)
+      ### when allow_partial_assignment is false, removes all affected visits of [service_id] because we can not affect all visits ###
+      @services_assignment[service_id][:vehicles].each{ |vehicle_id|
+        @services_assignment[service_id][:days].each{ |day|
+          route_data = @candidate_routes[vehicle_id][day]
+          removed_index = route_data[:stops].find_index{ |stop| stop[:id] == service_id }
+          next unless removed_index
+
+          service_point = route_data[:stops][removed_index][:point_id]
+          route_data[:stops].slice!(removed_index)
+          @services_data[service_id][:capacity].each{ |need, qty| route_data[:capacity_left][need] += qty }
+
+          unless route_data[:stops].any?{ |s| s[:point_id] == service_point }
```

Comment:
A formulation with `if` is easier to understand
```suggestion
          if route_data[:stops].none?{ |s| s[:point_id] == service_point }
```